### ISSUE TITLE
Implement CTC kernel

### ISSUE THREAD

**Issue by sherjilozair** [Theano/Theano#3338]
Current implementations of CTC in theano are quite slow as compared to native implementations. CTC being an important and fundamental cost function, it would be useful to have a fast CTC Op with GPU code.

**Comment by nouiz** [comment]
This can be useful. The most time consuming part is the GPU code itself. Do
you know any such implementation? I know one for LSTM that could be
relatively easily changed to make GRU.

I know nothing about CTC, is it similar to LSTM? From the difficulty people
had to make it work, I think it isn't very similar.

**Comment by sherjilozair** [comment]
Hi Frederic,

CTC is not in the same family as LSTM or GRU. CTC stands for connectionist
temporal classification, and is more of a cost function (thus, closer in
meaning to cross-entropy) than LSTM, GRU or RNN.

The uniqueness of CTC is that it is a cost function which can take in a
softmax over a sequence of length M, and a target of length N, where M
could be greater than N. In a sense, it is s generalization of
cross-entropy to the case when M > N. Obviously, this is an
over-simplification.

There is theano code for this out there:
1. https://github.com/shawntan/theano-ctc/blob/master/ctc.py
2. https://github.com/rakeshvar/rnn_ctc/blob/master/ctc.py
3. https://github.com/kastnerkyle/minet/blob/master/minet/costs.py

But they use Theano's diff engine to compute the gradient, and do not take
advantage of the analytic form available in the original paper (
ftp://ftp.idsia.ch/pub/juergen/icml2006.pdf).

I actually don't have enough GPU programming experience to do this myself.
However, I am trying an implementation with arrayfire.

On a side note, are there any plans in using arrayfire for GPU programming?
Is that even possible?

Thanks,
Sherjil Ozair

**Comment by nouiz** [comment]
If you get it working on GPU with ArrayFire, we can port that to Theano.

I'm not sure we can reuse ArrayFire in Theano. We need to check the license
and depending of it, maybe we need to rewrite it to not use it or we could
just include some of ArrayFire files in Theano itself.

I'm not sure I like that, but we could do that in a separate repo from the
main one.

Anyway, the first step is to get one version working.

**Comment by lamblin** [comment]
Discussion continued in #3871
PR at #5949
